### PR TITLE
Time out env

### PR DESCRIPTION
**Pull request by jeromedejaegher**
Les tests passent à 30 ms (précision du time.sleep()... Réouverture sur la branche dev_1.8.2 suite à la pertinente remarque de @laure-crochepierre !

### CODE REVIEW

**Comment by laure-crochepierre** [comment]
Il me samble qu'il te manque les commits du 17 avril qui corrigent les tests en question. =>  il faut faire un
```git pull upstream dev_1.8.2.``` si c'est bien ça

**Comment by BDonnot** [comment]
Ah oui je ne merge pas si les tests ne passe pas :-/

Et en plus il y a des conflits, à résoudre avant le merge également.

**Comment by BDonnot** [comment]
Also apparently `time.sleep()` is accurate up to 10 ms so 30 ms should be more than fine. The bug might be somewhere else.

File: grid2op/tests/test_timeOutEnvironment.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    def setUp(self) -> None:
+        super().setUp()
+        self.env0 = self.env1
+        self.env1 = self.env0.copy()
+        
+
+# TODO test runner
+# TODO test when used in gym (GymEnv)
+# TODO test that the "obs.simulate" and obs.get_forecast_env does not "do nothing"
+# TODO think about other tests
```
Comment (BDonnot) [inline grid2op/tests/test_timeOutEnvironment.py:146]: These tests need to be implemented

File: grid2op/tests/test_timeOutEnvironment.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import grid2op
+from grid2op.Environment import TimedOutEnvironment
+from grid2op.Agent import BaseAgent
+
+# Nota : time.sleep vs time.perf_counter() seems precise up to approx. 30 ms.
+
+class AgentOK(BaseAgent):
+    def __init__(self, env):
+        super().__init__(env.action_space)
+        self.time_out_ms = 0.7 * env.time_out_ms
```
Comment (BDonnot) [inline grid2op/tests/test_timeOutEnvironment.py:22]: can we keep 0.8 or 0.9 here ?

File: grid2op/tests/test_timeOutEnvironment.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        
+    def act(self, obs, reward, done):
+        time.sleep(1e-3 * self.time_out_ms)
+        return self.action_space()
+    
+    
+class AgentKO(BaseAgent):
+    def __init__(self, env):
+        super().__init__(env.action_space)
+        self.time_out_ms = 1.3 * env.time_out_ms
```
Comment (BDonnot) [inline grid2op/tests/test_timeOutEnvironment.py:32]: can we keep 1.1 or 1.2 here ?

File: grid2op/tests/test_timeOutEnvironment.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        
+    def act(self, obs, reward, done):
+        time.sleep(1e-3 * self.time_out_ms)
+        return self.action_space()
+    
+    
+class AgentKO1(BaseAgent):
+    def __init__(self, env):
+        super().__init__(env.action_space)
+        self.time_out_ms = 1.7 * env.time_out_ms
```
Comment (BDonnot) [inline grid2op/tests/test_timeOutEnvironment.py:42]: can we keep 1.8 or 1.9 here ?

File: grid2op/tests/test_timeOutEnvironment.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        
+    def act(self, obs, reward, done):
+        time.sleep(1e-3 * self.time_out_ms)
+        return self.action_space()
+            
+            
+class AgentKO2(BaseAgent):
+    def __init__(self, env):
+        super().__init__(env.action_space)
+        self.time_out_ms = 2.3 * env.time_out_ms
```
Comment (BDonnot) [inline grid2op/tests/test_timeOutEnvironment.py:52]: can we keep 2.1 or 2.2 here ?

File: grid2op/tests/test_timeOutEnvironment.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            act = agent.act(None, None, None)
+            obs, reward, done, info = self.env1.step(act)
+            assert info["nb_do_nothing"] == i % 3
+            assert info["nb_do_nothing_made"] == i % 3
+            assert self.env1._nb_dn_last == i % 3
+        
+
+class TestTimedOutEnvironment50(TestTimedOutEnvironment100):
+    def get_timeout_ms(self):
+        return 50
```
Comment (BDonnot) [inline grid2op/tests/test_timeOutEnvironment.py:133]: make that 150 to make the test pass due to the limitation of the "sleep" function

**Review by BDonnot — CHANGES_REQUESTED** [review]
Some other tests need to be implemented

File: grid2op/tests/test_timeOutEnvironment.py (hunk 1 above)
Comment (jeromedejaegher) [reply grid2op/tests/test_timeOutEnvironment.py:146]: needs further explanation of what is asked, what is gym, the link between obs and action space... We will see this after my holidays, in May !  😎 😝

**Pull request closed without merging**